File: 프로그래머스/2/12981. 영어 끝말잇기/영어 끝말잇기.cpp
#include <string>
#include <vector>
#include <iostream>
#include <unordered_set>

using namespace std;

vector<int> solution(int n, vector<string> words) {
    vector<int> answer;
    unordered_set<string> words_set;
    int idx = 1;
    int cnt = 0;
    string prev = "";
    for(auto& word : words) {
        
        if(cnt == 0){
            prev = word;
            words_set.insert(word);
            cnt++;
            idx++;
            continue;
        }
        
        // 탈락 조건3
        if(word[0] != prev[prev.size() - 1]){
            cnt /= n;
            cnt++;
            break;
        }
        
        // 탈락 조건 4
        if(words_set.find(word) != words_set.end()){
            cnt /= n;
            cnt++;
            break;
        }
        
        // 성공했을 시
        idx++;
        cnt++;
        prev = word;
        words_set.insert(word);
        
        if(idx > n)
            idx = 1;
        
        
        if(cnt == words.size()){
            idx = 0;
            cnt = 0;
        }
    }
    
    answer.push_back(idx);
    answer.push_back(cnt);

    return answer;
}

File: 프로그래머스/2/12981. 영어 끝말잇기/영어 끝말잇기.java
import java.util.*;
class Solution {
    public int[] solution(int n, String[] words) {
        Set<String> set = new HashSet<>();
        char prev = '1';
        
        for(int i = 0 ; i < words.length ; i++) {
            boolean isPossible = true;
            // 3번 예외사항
            if(words[i].charAt(0) != prev && i > 0){
                isPossible = false;
            }
            // 4번 예외사항
            if(set.contains(words[i])){
                isPossible = false;
            }
            // 5번 예외사항
            if(words[i].length() <= 1){
                isPossible = false;
            }
            
            if(!isPossible)
                return new int[]{(i % n) + 1, (i / n) + 1};
            
            prev = words[i].charAt(words[i].length() - 1);
            set.add(words[i]);
        }



        return new int[]{0, 0};
    }
}